replay2/memory_remapper/tests/memory_remapper_tests.cc

```diff
@@ -0,0 +1,205 @@
+	ASSERT_MOD_REPLAY_ADDRESS(remapper, captureAddress, replayAddress, size);
+
+	const ReplayAddress replayAddress2 = remapper.RemapCaptureAddress(captureAddress);
+	EXPECT_EQ(replayAddress, replayAddress2);
+
+	EXPECT_NO_THROW(remapper.RemoveMapping(captureAddress));
+	EXPECT_THROW(remapper.RemapCaptureAddress(captureAddress), MemoryRemapper::AddressNotMappedException);
+}
+
+TEST(MemoryRemapperTests, UnknownMapping) {
+
+	const size_t size = 128;
+
+	std::byte* rawCapturePtr = new std::byte[size];
+	CaptureAddress captureAddress(rawCapturePtr);
+
+	MemoryRemapper remapper;
+
+	EXPECT_THROW(remapper.RemapCaptureAddress(captureAddress), MemoryRemapper::AddressNotMappedException);
+
+	EXPECT_THROW(remapper.RemoveMapping(captureAddress), MemoryRemapper::AddressNotMappedException);
+	EXPECT_THROW(remapper.RemapCaptureAddress(captureAddress), MemoryRemapper::AddressNotMappedException);
+}
+
+TEST(MemoryRemapperTests, ZeroLengthMapping) {
+
+	const size_t size = 0;
+
+	std::byte* rawCapturePtr = new std::byte[size];
+	CaptureAddress captureAddress(rawCapturePtr);
+
+	MemoryRemapper remapper;
+	const MemoryObservation captureObservation(captureAddress, std::make_shared<ModResourceGenerator>(size));
+
+	EXPECT_THROW(remapper.AddMapping(captureObservation), MemoryRemapper::CannotMapZeroLengthAddressRange);
+	EXPECT_THROW(remapper.RemapCaptureAddress(captureAddress), MemoryRemapper::AddressNotMappedException);
+	EXPECT_THROW(remapper.RemoveMapping(captureAddress), MemoryRemapper::AddressNotMappedException);
+}
+
+TEST(MemoryRemapperTests, MultipleMappings) {
+
+	std::vector<ReplayAddress> replayAddresses;
+	std::vector<CaptureAddress> captureAddresses;
+
+	MemoryRemapper remapper;
+
+	for(int i = 0; i < 64; ++i) {
+
+		const size_t size = (i +1) * 2;
+
+		std::byte* rawCapturePtr = new std::byte[size];
+		CaptureAddress captureAddress(rawCapturePtr);
+		captureAddresses.push_back(captureAddress);
+
+		const MemoryObservation captureObservation(captureAddress, std::make_shared<ConstResourceGenerator>(std::byte(i), size));
+
+		const ReplayAddress replayAddress = remapper.AddMapping(captureObservation);
+		replayAddresses.push_back(replayAddress);
+
+		const ReplayAddress replayAddress2 = remapper.RemapCaptureAddress(captureAddress);
+		EXPECT_EQ(replayAddress, replayAddress2);
+	}
+
+	for(int i = 0; i < 64; ++i) {
+		const size_t size = i * 2;
+		EXPECT_NE(replayAddresses[i].bytePtr(), nullptr);
+		ASSERT_CONST_REPLAY_ADDRESS(remapper, captureAddresses[i], replayAddresses[i], std::byte(i), size);
+		EXPECT_NO_THROW(remapper.RemoveMapping(captureAddresses[i]));
+		EXPECT_THROW(remapper.RemapCaptureAddress(captureAddresses[i]), MemoryRemapper::AddressNotMappedException);
+	}
+}
+
+TEST(MemoryRemapperTests, MappingCollision) {
+
+	const size_t offset = 31;
+
+	const size_t sizeA = 128;
+	std::byte* rawCapturePtrA = new std::byte[sizeA];
+	CaptureAddress captureAddressA(rawCapturePtrA);
+
+	const size_t sizeB = sizeA -offset;
```

Comment:
So this is a test and this particular case is not an issue but this could overflow very easily and the exposing bug could be very cryptic.

One of the things that I don't like in C/C++ is that size_t is unsigned. I think that's one of the thing that Go has done right.

We should either use a signed length(int64) or we should be **VERY* cautious about using `-` operation.

What do you think about it?

Reply:
This just a test. I wrote it like this to use fewer magic numbers. I could have written 97 here either, but it's harder to notice 97 = 128-31

Reply:
My question was more on the we will probably do some operations with size and offsite and they being unsigned is something error prone. So I was asking should we change it or should we trust people that they will always check if offset is smaller than size.